Fix bug GHA "Schedule Thursday 1100"  (inactive members) 4768

t-will-gillis opened the pull request:
Fixes #4768 

### What changes did you make?
  - Since the current `schedule-thu-1100.yml` was scheduled to run on the first Thursday of the month, this action has been revised/ renamed to `schedule-monthly.yml` and now is scheduled to run on the first day of the month.
  - Because of this change, there no longer is a `schedule-thu-1100.yml` file.
  - Note that there is an existing action (#4788)  with the same name. This action/ file is renamed `schedule-monthly-PREV.yml`. This action is not/ has not been running either. The renaming is intended to be temporary because the plan is to explore how to integrate the other action's steps into this action. 
  - The "Get current date" step in the .yml file was deleted because a.) the code is deprecated and b.) it is unnecessary. (The previous file used this step to cause the automation to run only on the first Thursday of the month.)
  - The `working-directory` in the step "Install npm dependencies" is refactored to the correct directory i.e. `./github-actions/trigger-schedule/github-data`. (This was one of the reasons the existing GHA has not worked- this refactor fixes the action.)
  - The `node-version` is updated to version 18. (The previous version 14 is incompatible with the current modules.)
  - The `run` path in the step "Trim members" is refactored to the correct directory and the `contributors-data.js` is moved to the same directory as above. (This was another one of the reasons the existing GHA was not working. The `@octokit/rest` module cannot be found unless the js file resides in the same directory as the `package.json` file.) 
  - Since this js file is moved, the `trim-contributors` directory is empty and so is deleted.
  - Additional functionality has been added to `contributors-data.js`:
    - The automatic removal of members now occurs after two months of inactivity. 
    - Members who are one month inactive are placed on list so that they can be contacted. _(For a future PR, an automatic notification should be sent to these members to give them a chance to become active again before deletion= See **"FUTURE revisions/ additions"** below)_
  - Since files have been renamed and moved, the "Files changed" link above will not give a useful comparison of what is different. See the 'Diffchecker' screenshots below.

### Why did you make the changes (we will use this info to test)?
  - Changes were made primarily to get this action working again (it has been failing for over a year due to misdirected file paths.)
  - Other reasons are given in the text above.

### FUTURE revisions/ additions are detailed in ER #5163 

### Screenshots of Proposed Changes Of The Website 
<details>
<summary>Diffchecker for new `schedule-monthly.yml`</summary>

![Screenshot 2023-08-07 at 14-53-59 Computed Diff - Diff Checker](https://github.com/hackforla/website/assets/40799239/d18c1da4-0478-4039-a6f5-3f68c251434d)

</details>

<details>
<summary>Diffchecker for `contributors-data.js`</summary>

![Screenshot 2023-08-07 at 14-20-30 contributors-data js - Diff Checker](https://github.com/hackforla/website/assets/40799239/9313d4ca-c81f-4323-81d1-ac197f8e54ee)
![Screenshot 2023-08-07 at 14-20-50 contributors-data js - Diff Checker](https://github.com/hackforla/website/assets/40799239/dbde0bf7-1655-4e2a-9ec5-91675c06ac6c)
![Screenshot 2023-08-07 at 14-21-10 contributors-data js - Diff Checker](https://github.com/hackforla/website/assets/40799239/dcd2b689-11d3-4df0-9b9b-949171eecc06)
![Screenshot 2023-08-07 at 14-21-29 contributors-data js - Diff Checker](https://github.com/hackforla/website/assets/40799239/3d7d1c74-ec94-4cda-903c-96cab804c9dc)
![Screenshot 2023-08-07 at 14-21-43 contributors-data js - Diff Checker](https://github.com/hackforla/website/assets/40799239/93f5566c-e17b-4c48-a438-48f07902645d)
![Screenshot 2023-08-07 at 14-22-48 contributors-data js - Diff Checker](https://github.com/hackforla/website/assets/40799239/27333f7a-5a9f-4a46-8814-d9cae7969c61)

</details>

- Here is a [test run of the GHA](https://github.com/t-will-gillis/website/actions/runs/5790131810/job/15692598438) from my repo.

t-will-gillis committed:
Update and rename schedule-monthly.yml to schedule-monthly-PREV.yml

t-will-gillis committed:
Update and rename schedule-thu-1100.yml to schedule-monthly.yml

t-will-gillis committed:
Create contributors-data.js

t-will-gillis committed:
Delete github-actions/trigger-schedule/trim-contributors directory

t-will-gillis committed:
Update contributors-data.js

t-will-gillis committed:
Update schedule-monthly.yml

t-will-gillis committed:
Update contributors-data.js

add misc. explicit semicolons

LRenDO commented:
Hi @mademarc! When you have a minute, please add your ETA and Availablity.  Thanks!

mademarc commented:
Review ETA: 8/11/2023
Availability: 5:11PM

mademarc reviewed (dismissed):
Hey @t-will-gillis  the format also the code changes on the line 5 & 8 on the yml file is done correctly   and @LRenDO the ETA and Availability is done.

one2code commented:
**ETA:** 8/12/23

one2code commented:
Just wanted to comment and say that I have been reviewing this pull request, but due to the complexity and amount of file changes / line changes, I may have to split this one into chunks. I know the challenges that we have been facing with the various schedule files due to things such as deprecation, breaking changes on updates, or changes to how the schedules and triggers need to be setup in the latest updates to avoid breaking functionality. With that said, I want to take my time with this and make sure I can verify under testing the latest changes, with the understanding that this is one key step in solving several other related issues. Likewise, if I were to miss something in testing, then that would create a domino effect, making the other related issues difficult to solve for.

Thank you for your patience as I continue to delve deep on this.

LRenDO commented:
Hi @yujioshiro! Thanks for volunteering to review this issue! When you have a minute, please add your ETA and Availability. Thanks!

LRenDO commented on github-actions/trigger-schedule/github-data/contributors-data.js:
```diff
@@ -0,0 +1,247 @@
+// inactive, and at two months remove contributor from team(s)
+const today = new Date();
+let oneMonthAgo = new Date(today.setMonth(today.getMonth() - 1)); // 1 month from today
+let twoMonthsAgo = new Date(today.setMonth(today.getMonth() - 1)); // 1 month behind oneMonthAgo
```
Hi @t-will-gillis! I went to review this and I believe I don't have correct permissions to access HFLA Org level information using the API.  I get back the repo level information (in `fetchContributors`) just fine when I run the test, but the org level information (in `fetchTeamMembers`) gives a 404 response which seems like it is likely a permissions error. See [GitHub doc](https://docs.github.com/en/get-started/learning-about-github/access-permissions-on-github). Did you run into this issue at all when running the GHA manually? I was wondering if your maintainer role grants the right permissions.

That said, I did read through the code and PR and it looks like you did a great job refactoring and solving the bug! I particularly appreciated the explanation and breakdown of what needed to be done to fix the issue in your PR. The inclusion of the diffs was very helpful as well since many things were moved and deleted. It was also useful to see the ER with the other issues that need to be addressed including adding the notify functionality and the other monthly actions to this GHA.

I have a couple additional thoughts/questions based on what I read in your PR and code.
- I did noticed something here on line 17.  It looks like `twoMonthsAgo` is set to the same date as `oneMonthAgo`.  I might be missing something since I haven't been able to thoroughly test it, but I believe it should be `- 2` and not `- 1` at the end of line 17. 
- I believe also that the comment in line 229 in `notifyInactiveMembers` should say `oneMonthAgo` instead of `twoMonthsAgo`. The code itself looks correct for that function though since you're passing `contributorsOneMonthAgo` in on line 43.
- In the first mention of `<< FUTURE` in the PR, I wasn't sure exactly what the future part was referring to.  It looks like the files and directories were already moved and deleted. Was there something else that it was referring to that will be done in the future that I missed?

If/when we can get the permissions issue resolved, I will circle back.

LRenDO commented:
Hi @t-will-gillis! I went to review this and I believe I don't have correct permissions to access HFLA Org level information using the API.  I get back the repo level information (in `fetchContributors`) just fine when I run the test, but the org level information (in `fetchTeamMembers`) gives a 404 response which seems like it is likely a permissions error. See [GitHub doc](https://docs.github.com/en/get-started/learning-about-github/access-permissions-on-github). Did you run into this issue at all when running the GHA manually? I was wondering if your maintainer role grants the right permissions.

That said, I did read through the code and PR and it looks like you did a great job refactoring and solving the bug! I particularly appreciated the explanation and breakdown of what needed to be done to fix the issue in your PR. The inclusion of the diffs was very helpful as well since many things were moved and deleted. It was also useful to see the ER with the other issues that need to be addressed including adding the notify functionality and the other monthly actions to this GHA.

I have a couple additional thoughts/questions based on what I read in your PR and code.
- I did noticed something on line 17 in [contributors-data.js](https://github.com/hackforla/website/pull/5158/files#diff-2cd4ed23c62a269fe57e80b69f61b503ea5f7721c0fbbfbf2d4a000069c9457e).  It looks like `twoMonthsAgo` is set to the same date as `oneMonthAgo`.  I might be missing something since I haven't been able to thoroughly test it, but I believe it should be `- 2` and not `- 1` at the end of line 17. 
- I believe also in [contributors-data.js](https://github.com/hackforla/website/pull/5158/files#diff-2cd4ed23c62a269fe57e80b69f61b503ea5f7721c0fbbfbf2d4a000069c9457e) that the comment in line 229 in `notifyInactiveMembers` should say `oneMonthAgo` instead of `twoMonthsAgo`. The code itself looks correct for that function though since you're passing `contributorsOneMonthAgo` in on line 43.
- In the first mention of `<< FUTURE` in the PR, I wasn't sure exactly what the future part was referring to.  It looks like the files and directories were already moved and deleted. Was there something else that it was referring to that will be done in the future that I missed?

If/when we can get the permissions issue resolved, I will circle back.

LRenDO commented on github-actions/trigger-schedule/github-data/contributors-data.js:
```diff
@@ -0,0 +1,247 @@
+// inactive, and at two months remove contributor from team(s)
+const today = new Date();
+let oneMonthAgo = new Date(today.setMonth(today.getMonth() - 1)); // 1 month from today
+let twoMonthsAgo = new Date(today.setMonth(today.getMonth() - 1)); // 1 month behind oneMonthAgo
```
Here's where it looks like `twoMonthsAgo` is set to the same date as `oneMonthAgo`

t-will-gillis replied:
Although the code on lines 15-17 do set the correct dates for `oneMonthAgo` and `twoMonthsAgo`,*** it is confusing and not at all apparent. Please see the revision. 

Thank you for flagging this- it needed revising.

***In the end, the dates are set correctly because the value of "today" on line 15 shifts one month prior on 16 and one additional month prior (two total) on line 17.

LRenDO commented on github-actions/trigger-schedule/github-data/contributors-data.js:
```diff
@@ -0,0 +1,247 @@
+
+
+/**
+ * Function to return list of contributors that have been inactive since twoMonthsAgo
```
Here's where I believe it should say `oneMonthAgo` instead of `twoMonthsAgo`

t-will-gillis replied:
Good catch! Yes, this should say `oneMonthAgo`. I will make this change on the commit.

t-will-gillis committed:
Update contributors-data.js

Updated definitions for `oneMonthAgo` and `twoMonthsAgo`, and corrected reference to "one month ago" in comment at function `notifyInactiveMembers`

t-will-gillis commented:
Hey @LRenDO! Thank you for your thorough review and comments. 
- Regarding the 404 response: I did not run into this when I was testing. Since I believe we have similar roles on the website teams, I am wondering if the problem could be due to our respective tokens having different scopes/ permissions? For comparison, my token scopes are: ‘admin:org’, ’project’, ’repo’, and ‘workflow’. 
- Regarding lines 15-17 in `contributors-data.js`: Thank you for flagging this. Although this code does work as intended (i.e., `oneMonthAgo` and `twoMonthsAgo` are actually set to the correct dates one month and two months in the past), it is confusing. Please see the revision.
- Regarding line 229 in `contributors-data.js`: Good catch! Yes, the comment should say “oneMonthAgo”
- I edited the description referring to the "<< FUTURE " revisions.

Let me know, and I would be happy to demo this if it would be helpful-

LRenDO commented:
Hey @t-will-gillis! 

On Lines 15-17, I was reading through too quickly. I see now that today was also being set to one month ago which made it update correctly in the code. What you've got now is definitely clearer though. The PR and comment look great too! Thanks for taking the time to review and clarify what I was seeing!

On testing, the permissions seem to cover those areas. I'll connect with you on Slack and maybe you can catch something I'm missing.

LRenDO reviewed (approved):
Hi @t-will-gillis!

I was able to figure out that a classic PAT is required for this GHA. I had been using a Fine-grained PAT. The testing went as expected and the changes all look great here! 

I see after the issue is merged, a wiki page will be created for the GHA. If the notification implementation will not be completed in the very near future, I would mention there and/or in the test logs that the notification feature has not yet been implemented and that the logs are providing a list of folks who will be notified when the implementation is complete.  Additionally, in the wiki I would make sure to include that it does not currently remove 60 day inactive members from child teams or the website team, only specifically the website-write team.

Thanks for taking the time to contribute to fixing this GHA! The GHAs can definitely be pretty cumbersome!

LRenDO commented:
Hi @one2code and @yujioshiro!
I thought I would pass along how I was able to run this GHA successfully in case it's helpful for y'all.

- [ ] Add a **classic** token with the following permissions `admin:org`, `project`, `repo`, and `workflow`
[How to create a classic personal access token](https://docs.github.com/en/authentication/keeping-your-account-and-data-secure/managing-your-personal-access-tokens#creating-a-personal-access-token-classic)
- [ ] Copy the token
- [ ] Create a repository action secret
[How to create a repository action secret](https://docs.github.com/en/actions/security-guides/encrypted-secrets#creating-encrypted-secrets-for-a-repository)
- [ ] Create a branch locally and pull this feature branch as you would usually do for testing
- [ ] Make the following changes in `schedule-monthly.yml`
  - [ ] Update `'hackforla/website'` to your forked repo path
  - [ ] Update `GITHUB_TOKEN` to the name of your repository action secret you created in the third step
  - [ ] Update `HACKFORLA_BOT_PA_TOKEN` to the name of your repository action secret you created in the third step
  - [ ] Add `workflow dispatch:` after `on:` at the same indention level as `schedule`
  [How to set up a workflow for manual testing](https://docs.github.com/en/actions/using-workflows/manually-running-a-workflow#configuring-a-workflow-to-run-manually)
- [ ] Push the branch to your forked repo
- [ ] Change the default branch on your repo to the branch you pushed
[How to change the default branch](https://docs.github.com/en/repositories/configuring-branches-and-merges-in-your-repository/managing-branches-in-your-repository/changing-the-default-branch#about-changing-the-default-branch)
- [ ] Run the workflow manually
[How to run a workflow manually](https://docs.github.com/en/actions/using-workflows/manually-running-a-workflow#running-a-workflow)

yujioshiro commented:
I will start to work on reviewing this PR tomorrow. Review ETA: 9/1/2023

t-will-gillis commented:
Hi @yujioshiro would you please provide an update on your progress and let us know if you are still looking at this? Also, please ask if you would like something specific explained. This is a time sensitive issue and needs to be resolved so that other issues can proceed. Thank you
